src/DGtal/base/SetFunctions.h

```diff
@@ -0,0 +1,1469 @@
+      {
+        return EqualPredicate( less( C ) );
+      }
+      static inline const key_type& key( const value_type& value )
+      {
+        return value.first;
+      }
+    };
+
+    /// unordered set-like adapter.
+    template <typename Container>
+    struct ComparatorAdapter< Container, true, false, false > 
+    {
+      typedef typename Container::value_type  value_type;
+      typedef typename Container::key_type    key_type;
+      typedef std::less< key_type >           LessThanPredicate;
+      typedef std::equal_to< key_type >       EqualPredicate;
+      static LessThanPredicate less( const Container& /* C */ )
+      {
+        return LessThanPredicate();
+      }
+      static EqualPredicate equal_to( const Container& /* C */ )
+      {
+        return EqualPredicate();
+      }
+      static inline const key_type& key( const value_type& value )
+      {
+        return value;
+      }
+    };
+
+    /// unordered map-like adapter.
+    template <typename Container>
+    struct ComparatorAdapter< Container, true, false, true > 
+    {
+      typedef typename Container::value_type      value_type;
+      typedef typename Container::key_type        key_type;
+      typedef KeyComparatorForPairKeyData
+        < std::less< key_type >, value_type >     LessThanPredicate;
+      typedef KeyComparatorForPairKeyData
+        < std::equal_to< key_type >, value_type > EqualPredicate;
+
+      static LessThanPredicate less( const Container& /* C */ )
+      {
+        return LessThanPredicate( std::less< key_type >() );
+      }
+      static EqualPredicate equal_to( const Container& /* C */ )
+      {
+        return EqualPredicate( std::equal_to< key_type >() );
+      }
+      static inline const key_type& key( const value_type& value )
+      {
+        return value.first;
+      }
+
+    };
+
+
+
+    /**
+     * Description of template class 'SetFunctions' <p> \brief Aim:
+     * Specialize set operations (union, intersection, difference,
+     * symmetric_difference) according to the given type of
+     * container. It uses standard algorithms when containers are
+     * ordered, otherwise it provides a default implementation.
+     *
+     * @tparam Container any type of container.
+     *
+     * @tparam associative tells if the container is associative
+     * (e.g. set, map, unordered_set, unordered_map).
+     *
+     * @tparam ordered tells if the container is ordered (e.g., set, map).
+     *
+     *
+     * Specialized implementations are marked with (S) in the list below.
+     *
+     * |--------------------|-------------|----------|----------|
+     * | Container          | associative | ordered  |  pair    |
+     * |--------------------|-------------|----------|----------|
+     * | vector             |   false     |  false   |  false   |
+     * | list               |   false     |  false   |  false   |
+     * | (to avoid)         |   false     |  false   |  true    |
```

Comment:
In case somebody use a weird container for set: non associative, non sorted, and with a pair key/value. Like an unsorted list of pairs. Not a reasonable choice for a set structure.